connector/src/main/java/com/datastax/oss/pulsar/source/converters/NativeJsonConverter.java

```diff
@@ -0,0 +1,407 @@
+            DataType dataType = udtValue.getType(field);
+            if (fields.contains(field.asInternal()) && !udtValue.isNull(field)) {
+                node.set(field.toString(), toJson(udtValue, field, dataType));
+            }
+        }
+
+        return node;
+    }
+
+    JsonNode createArrayNode(org.apache.avro.Schema schema, Collection collection) {
+        ArrayNode node = JsonNodeFactory.instance.arrayNode(collection.size());
+        for(Object element : collection) {
+            if (element instanceof UdtValue) {
+                UdtValue udtValue = (UdtValue) element;
+                node.add(buildUDTValue(udtValue));
+            } else if (element instanceof Set) {
+                node.add(createArrayNode(schema.getElementType(), (Collection)element));
+            } else if (element instanceof List) {
+                node.add(createArrayNode(schema.getElementType(), (Collection)element));
+            } else {
+                JsonNode fieldValue = toJson(schema.getElementType(), element);
+                node.add(fieldValue);
+            }
+        }
+
+        return node;
+    }
+
+    public static byte[] serializeJsonNode(JsonNode node) {
+        try {
+            return mapper.writeValueAsBytes(node);
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public byte[] fromConnectData(GenericRecord genericRecord) {
+        if (genericRecord == null) {
+            return null;
+        }
+        ObjectNode objectNode = jsonNodeFactory.objectNode();
+        for (org.apache.avro.Schema.Field field : genericRecord.getSchema().getFields()) {
+            objectNode.set(field.name(), toJson(field.schema(), genericRecord.get(field.name())));
+        }
+        return serializeJsonNode(objectNode);
+    }
+
+    private static Map<String, LogicalTypeConverter<?>> logicalTypeConverters = new HashMap<>();
+
+    private static JsonNode cqlDurationToJsonNode(CqlDuration cqlDuration) {
+        ObjectNode object = jsonNodeFactory.objectNode();
+        object.put(CqlLogicalTypes.CQL_DURATION_MONTHS, cqlDuration.getMonths());
+        object.put(CqlLogicalTypes.CQL_DURATION_DAYS, cqlDuration.getDays());
+        object.put(CqlLogicalTypes.CQL_DURATION_NANOSECONDS, cqlDuration.getNanoseconds());
+        return object;
+    }
+
+    private static JsonNode bigDecimalToJsonNode(BigDecimal bigDecimal) {
+        ObjectNode decimalNode = jsonNodeFactory.objectNode();
+        decimalNode.put(CqlLogicalTypes.CQL_DECIMAL_BIGINT, bigDecimal.unscaledValue().toByteArray())
+                .put(CqlLogicalTypes.CQL_DECIMAL_SCALE, bigDecimal.scale());
+
+        return decimalNode;
+    }
+    public static JsonNode toJson(org.apache.avro.Schema schema, Object value) {
+        // schema.getLogicalType() always returns null although the name would be populated with the logical type
+        // TODO: Use logical type instead of name
```

Comment:
we should track a issue

Reply:
added in the new patch: https://github.com/datastax/cdc-apache-cassandra/issues/85